### DISCUSSION TITLE
Partial results with graphcache

### DISCUSSION THREAD

**Discussion by marcel-happyfloat** [urql-graphql/urql#1834] [Q&A] [answered]
Sorry, but I'm still lost on this topic with URQL.

By now, I understand that it will not resolve partial data if some other data is marked as non-null. 

- This by itself is already a problem for me because my schema is auto generated based on the database schema. The database schema is of course different to the multiple UI clients (e.g. app vs. admin-panel), so non-null in the database can still be null in the UI client. I think I can work around that by extending the generated schema before loading it into URQL.
- The other point is some totally unrelated query within the same GraphQL query will also prevent the partial resolve.

Example: You got a list of items:

```gql
query GetItems {
  content {
    id
    title
  }
}
```

In your Detail Page you have something like this:

```gql
query GetDetailPage {
  content_by_pk {
    id
    title
    some_optional_other_value
  }
  # this will prevent resolving my partial content data... why!?
  some_other_stuff {
    other_stuff
  }
}
```

Somehow the most basic use case of a normalized cache is driving me crazy :(
It's working with apollo, but it's so bloated and I try to stay with URQL.

Do I really have to split one request into multiple to solve this problem? Thanks for your effort on answering all these questions from the community!

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

Yeah, relying on `non-null` in the schema is generally a good idea as you've indicated that this field is optional, this extends to `type-generation` with for instance `grapqhl-code-generator` where nullable fields will also be nullable in your typings which makes it easier to create UI's that reflect the nullability constraints of your schema.

Data being "fully partial" without any way to deterministically tell whether it can or can't be there will sooner or later bite you with some error (or you'd have to guard for _every single property_). Hence this is our heuristic currently.

This extends to two `query-fields` being within the operation, if you want to have 1 you can split it out into two separate `useQuery` calls.

I hope this helps shed light on the reasoning behind this partial constraint.

**Reply by marcel-happyfloat** [reply]
Sorry, I just don't get it :(. Feel free to delete everything

**Reply by JoviDeCroock** [reply]
Which part don't you get I'm happy to elaborate further, no need to apologise for anything!

**Reply by marcel-happyfloat** [reply]
Ok thank you. Let me try again. I think my understanding changed already for the better.

1. Let's talk about "non-null" fields again. My latest understanding is, that non-null in the graphql schema means, that if I request the field, I will receive a non-null value. For me, this is totally unrelated to the UI and my auto generated schema is actually correct this way. 

2. Hopefully getting point 1. correct now, I think I understand the reasoning behind not resolving the partial stuff. But for me it feels not practical. How are you going to ever solve this standard use case of "ListItem -> DetailView". The DetailView will always request more non-null data in real world examples, so the already cached data will be obsolete?

3. As a workaround I can push the ListItem data to the new Screen and use them as initial/cached values, before overwriting them with the data from the DetailView query. Or I could create a second query, which only contains cached fields and execute it with cache-only policy and also overwrite it with the other query data afterwards. Both cases seem overly complicated instead of just populating the data partially with the cached values... maybe URQL could add another field `partialData` in the result of the query, so we do not mix the two and the typings stay enforced? So we could use it in a similar fashion as my current workarounds?

**Reply by JoviDeCroock** [reply]
Point 1 already kind off disconnects already a bit from GraphQL as we intend to provide a contract between UI and server which is different from REST where the application got a response and was just expected to deal with it. GraphQL makes UI requirements easier this way. For instance if we ask for a person:

```gql
query {
  persons {
    id
    firstName
    lastName
  }
}
```

We expect to get those fields, when a schema sais that `lastName` is nullable then we can assume the UI is also tailored for this not being present. When the schema sais the inverse we can assume that the UI is tailoring towards needing this.

Point 2 is not necessarily true, I can talk to a few examples that I've seen in the wild, where we have a product-list that needs basic-information like the pictures, ... for a product and when we go into the details-view we only need to fetch additional information like for instance additional product-information which was optional to have. I do get what you're saying where this is not nullable but it's hard to provide a better heuristic that doesn't introduce a footgun to any users. If we consider everything to be nullable without looking at the schema we make it impossible for devs to predict what dat will be present at what point in time. As often times it's much more complex than just going from list --> details, consider the case of going from a cart --> details, ... there are much more data entry points and often times much more complex. Hence we've taken a heuristic that is easily combined with something like `TypeScript` as tools use this same heuristic to say if something is nullable or not.

To point 3, I guess you could introduce two queries _or_ have your own exchange that applies the heuristics for your specific use case.

**Comment by marcel-happyfloat** [comment]
> intend to provide a contract between UI and server

That seems very odd to me. I thought GraphQL is exactly the opposite. You can have different Clients and different components, each with it's own requirements.

Anyway, I will now get my cached data from wherever and just merge it myself with the new query, would have been nice to be able to use the "partial data" feature for this.

Thanks!

**Reply by JoviDeCroock** [reply]
I mean there's possibly another solution, depending on what client you are using there are possible solutions. Generally in a system where we consider scalars/links it's dangerous to have everything be non-null as that accounts for a "perfect world" scenario, no errors can occur _or_ you'll never show data. Consider when a product has reviews, when we fetch the reviews as a child-field of product and the reviews fetch fails, we still want to be able to show the product. 

Basically marking these connections as mandatory has a lot of side-effects that could prove erroneous in the future, this could also prove hard to deprecate fields in the future. Tools like PostGraphile, ... often have escape hatches to bypass these constrains and mark something as nullable.>

> That seems very odd to me. I thought GraphQL is exactly the opposite. You can have different Clients and different components, each with it's own requirements.

Exactly but the properties in and of themselves aren't the issue here it's how data flows that seems to be the problem here, right?

**Reply by marcel-happyfloat** [reply]
I don't find a way to mark something as nullable in [Hasura](https://hasura.io/) unfortunately and I also prefer to have the same schema for the App and the Admin-Panel. The Admin panel will require more non-values of course.
